### ISSUE TITLE
EF Core 7 Rev eng: Missing HasForeignKey when referencing a navigation property with a unique constraint with DataAnnotations on

### ISSUE THREAD

**Issue by ErikEJ** [dotnet/efcore#29418]
## File a bug

Using EF Core 7 SQL Server reverse engineering with "Use Data annotations" enabled, the foreignkey configuration is missing.

See description and link to repro schema here: https://github.com/ErikEJ/EFCorePowerTools/issues/1554

@bricelam 

EF Core version: 7.0 RC2
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 6
Operating system: Windows

**Comment by bricelam** [comment]
@AndriySvyryd Is it by design that this model creates a new `Car.ColorCode1` property (instead of using the existing `ColorCode` property) for the relationship?

```cs
class TestContext : DbContext
{
    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Car>().HasOne(d => d.Color).WithMany(p => p.Cars)
            .HasPrincipalKey(p => p.ColorCode);
    }
}

class Car
{
    [Key]
    public int Id { get; set; }

    public string? ColorCode { get; set; }
    public virtual Color? Color { get; set; }
}

class Color
{
    [Key]
    public int Id { get; set; }

    public string ColorCode { get; set; } = null!;
    public virtual ICollection<Car> Cars { get; } = new List<Car>();
}
```

**Comment by AndriySvyryd** [comment]
Yes, we only match these patterns:
 * [navigation property name][principal key property name]
 * [navigation property name]Id
 * [principal entity name][principal key property name]
 * [principal entity name]Id

**Comment by ajcvickers** [comment]
@AndriySvyryd In that case, why do we create a foreign key name by convention (`ColorCode`--the principal key name) that by default isn't something we match? That seems inconsistent--if we don't match it, we shouldn't create the thing we don't match.

**Comment by TimKras** [comment]
I just wanted to add that this problem only occurs in EF Core 7 (RC).
It is working fine in EF Core 6.

**Comment by AndriySvyryd** [comment]
@ajcvickers Because naming it `ColorColorCode` looked wrong. However, using the same logic for FK discovery might be unexpected. But we could consider using it as long as it doesn't match PK properties.

**Comment by ajcvickers** [comment]
Notes from triage: ideally, the rules would not result in this behavior. However, rather than making the rules _even more complex_ we decided that it would be better here for scaffolding to respect the existing rules, as it did in EF Core 6.0.l

**Comment by ErikEJ** [comment]
No plan to fix for 7.0??

**Comment by ajcvickers** [comment]
@ErikEJ We are going to consider patching.

**Comment by bricelam** [comment]
Stranger and stranger...

Trying to configure it with an attribute doesn't work.

```cs
// Still not used. Still get shadow property 'ColorCode1'
public string? ColorCode { get; set; }

[ForeignKey(nameof(ColorCode))]
public virtual Color? Color { get; set; }
```

**Comment by ajcvickers** [comment]
What did 6.0 scaffold?

**Comment by bricelam** [comment]
Scaffold? It never scaffolded data annotations when the principal key is non-primary. It always used fluent API.

But I'm just playing with model building to see which data annotations we can scaffold. I'll just work around the `[ForeignKey]` bug for now, and we can address it separately.

**Comment by bricelam** [comment]
Filed #29448

**Comment by TimKras** [comment]
@ajcvickers will this be patched in version 7? If so, is it possible to give an estimation on the timing?

**Comment by ajcvickers** [comment]
@TimKras Possibly, although you are the only person to have reported it so far, so right now it might not meat the customer impact bar required for a servicing fix.

**Comment by ajcvickers** [comment]
Note from triage: this issue covers a wider range of cases than initially thought; we should patch.

**Comment by bricelam** [comment]
## 🩹 Workaround

I may be stating the obvious here, but one workaround is to not use data annotations when scaffolding. Or, you can manually add `HasForeignKey` to the scaffolded code after its generated.

**Comment by TimKras** [comment]
> ## 🩹 Workaround
> I may be stating the obvious here, but one workaround is to not use data annotations when scaffolding. Or, you can manually add `HasForeignKey` to the scaffolded code after its generated.

A better temporary workaround is adding the missing `HasForeignKey` in the OnModelCreatingPartial method, so it not overridden each time the context is scaffold: 

partial void OnModelCreatingPartial(ModelBuilder modelBuilder)
{
   modelBuilder.Entity<Car>().HasOne(d => d.Color).WithMany(p => p.Cars)
     .HasForeignKey(p=>p.ColorCode)
     .HasPrincipalKey(p => p.ColorCode);
}

**Comment by TimKras** [comment]
Confirmed fixed in 7.0.3
